jupyterhub/oauth/provider.py

```diff
@@ -558,19 +558,23 @@ def add_client(self, client_id, client_secret, redirect_uri, description=''):
 
         hash its client_secret before putting it in the database.
         """
-        # clear existing clients with same ID
-        for orm_client in self.db.query(orm.OAuthClient).filter_by(
-            identifier=client_id
-        ):
-            self.db.delete(orm_client)
-        self.db.commit()
-
-        orm_client = orm.OAuthClient(
-            identifier=client_id,
-            secret=hash_token(client_secret),
-            redirect_uri=redirect_uri,
-            description=description,
+        # Update client if it already exists, else create it
+        # Sqlalchemy doesn't have a good db agnostic UPSERT,
+        # so we do this manually. It's protected inside a
+        # transaction, so should fail if there are multiple
+        # rows with the same identifier.
+        # TODO: Is this safe?
```

Comment:
Yes, this is safe. This is the pattern we use to create ~everything (Users, groups, etc.).

We may need to come back to recovering more gracefully from failures due to races if and when we get to developing horizontally scalable Hubs, but it can't fail right now since the failure addressed by UPSERT can only occur with multiple db writers, which is impossible with jupyterhub today.